File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() == 0) return 0;
        int ans = 0;
        int low = 0;
        int high = height.size() - 1;
        int lMax = height[low];
        int rMax = height[high];
        while (low < high) {
            if (lMax < rMax) {
                low++;
                lMax = max(lMax, height[low]);
                ans += lMax - height[low];
            } else {
                high--;
                rMax = max(rMax, height[high]);
                ans += rMax - height[high];
            }
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        max_area = 0
        max_left = 0
        max_right = 0
        l = 0
        r = len(h) - 1

        while l < r:
            if h[l] < h[r]:
                max_left = max(max_left, h[l])
                if max_left > h[l]:
                    max_area += max_left - h[l]
                l += 1
            else:
                max_right = max(max_right, h[r])
                if max_right > h[r]:
                    max_area += max_right - h[r]
                r -= 1

        return max_area